NTPClient.cpp

```diff
@@ -90,40 +90,108 @@ bool NTPClient::forceUpdate() {
+    Serial.print(tok);Serial.print(", ");
+    Serial.println((tok-tik)/2.0);
+  #endif
 
   this->_udp->read(this->_packetBuffer, NTP_PACKET_SIZE);
-
-  unsigned long highWord = word(this->_packetBuffer[40], this->_packetBuffer[41]);
-  unsigned long lowWord = word(this->_packetBuffer[42], this->_packetBuffer[43]);
+  uint32_t high_word,low_word;
+  uint32_t receive_int,transmit_int; //integer part
   // combine the four bytes (two words) into a long integer
   // this is NTP time (seconds since Jan 1 1900):
-  unsigned long secsSince1900 = highWord << 16 | lowWord;
+  high_word=word(this->_packetBuffer[32],this->_packetBuffer[33]);
+  low_word=word(this->_packetBuffer[34],this->_packetBuffer[35]);
+  receive_int=(high_word<<16) | low_word;
+  high_word=word(this->_packetBuffer[40],this->_packetBuffer[41]);
+  low_word=word(this->_packetBuffer[42],this->_packetBuffer[43]);
+  transmit_int=(high_word<<16) | low_word;
+  #ifdef DEBUG_NTPClient
+    Serial.print("receive_int, transmit_int: ");
+    Serial.print(receive_int);Serial.print(", ");
+    Serial.println(transmit_int);
+  #endif
 
-  this->_currentEpoc = secsSince1900 - SEVENZYYEARS;
+  float receive_dec=0,transmit_dec=0; //decimal part
+  high_word=word(this->_packetBuffer[36],this->_packetBuffer[37]);
+  receive_dec=high_word/65536.0;
+  #ifdef DEBUG_NTPClient
+    Serial.print("receive_dec, transmit_dec: ");
+    Serial.print(high_word,HEX);Serial.print(", ");
+    Serial.print(receive_dec,6);Serial.print(", ");
+  #endif
+  high_word=word(this->_packetBuffer[44],this->_packetBuffer[45]);
+  transmit_dec=high_word/65536.0;
+  #ifdef DEBUG_NTPClient
+    Serial.print(high_word,HEX);Serial.print(", ");
+    Serial.println(transmit_dec,6);
+  #endif
+
+  float ping_delay;
+  ping_delay=(tok-tik)/1000.0-(transmit_int-receive_int)-(transmit_dec-receive_dec);
+  ping_delay/=2.0;
+  if(ping_delay<=0){
+    Serial.println("ERROR: ping_delay < 0.0!");
+  }
+
+  this->_lastUpdate=tok;
+  this->_currentEpoc=transmit_int - SEVENZYYEARS ;
+  this->_current_epoc_dec=ping_delay+transmit_dec;
+  if(this->_current_epoc_dec>1){
+    this->_currentEpoc+=(int)this->_current_epoc_dec;
+    this->_current_epoc_dec-=(int)this->_current_epoc_dec;
+  }
+
+  #ifdef DEBUG_NTPClient
+    Serial.print("current Epoc: ");
+    Serial.print(this->_currentEpoc);Serial.print(" ");
+    Serial.println(this->_current_epoc_dec,6);
+  #endif
 
   return true;  // return true after successful update
 }
 
-bool NTPClient::update() {
-  if ((millis() - this->_lastUpdate >= this->_updateInterval)     // Update after _updateInterval
-    || this->_lastUpdate == 0) {                                // Update if there was no update yet.
-    if (!this->_udpSetup || this->_port != NTP_DEFAULT_LOCAL_PORT) this->begin(this->_port); // setup the UDP client if needed
-    return this->forceUpdate();
+int8_t NTPClient::update() {
+  uint32_t now=millis();
+  if(now>=this->_lastUpdate){ //if not overflow
```

Comment:
I'm not sure you need this logic. `millis() - this->_lastUpdate` should always give a positive integer since millis and lastUpdate are both unsigned.

Reply:
millis() rollover (depending on mcu) every ~49 days